src/shred_collector/shred_receiver.zig

```diff
@@ -80,76 +80,69 @@ pub const ShredReceiver = struct {
     /// Returns when exit is set to true.
     fn runPacketHandler(
         self: *Self,
-        receivers: []const *Channel(ArrayList(Packet)),
-        response_sender: *Channel(ArrayList(Packet)),
+        receivers: []const *Channel(Packet),
+        response_sender: *Channel(Packet),
```

Comment:
The `ArrayList` of packets doesn't make much sense, since the channel already stores its items in a managed state. When we "send" a packet into the channel you can think of it as equivalent to appending to an `ArrayList`. The only difference is that we're using less memory by not storing the extra array list metadata in every item we send.

Reply:
> The only difference is that we're using less memory by not storing the extra array list metadata in every item we send.

There is another difference which I explained in my first comment. The arraylist approach reduces the amount of memory that needs to be copied.

Let's say we just received 100 packets. We have two options:

### The "arraylist" approach

This doesn't necessarily need to be an arraylist. I'll refer to it as a "pointer" to be more generic.

socket reader thread:
1. Allocate memory large enough to hold 100 packets
2. Write each packet into that region of memory
3. Write the pointer into a channel

shred receiver thread:

4. Read the pointer from the channel
5. Basic validation of each packet
6. Write the pointer into a channel

shred verifier thread:

7. Read the pointer from the channel
8. Verify the thread of each signature
9. Write the pointer into a channel

shred processor thread:

10. Read the pointer from the channel
11. Write each shreds into the blockstore

With this approach, steps 2 and 11 each copy 120 KB. The other steps all copy very little data in comparison.


### The pass-each-packet-by-value approach

<skip steps 1-2>

shred verifier thread:

3. Write each packet into the channel

shred receiver thread

4. Read each packet from the channel
5. Basic validation of each packet
6. Write the pointer to the buffer into a channel

shred verifier thread:

7. Read the pointer from the channel
8. Verify the thread of each signature
9. Write the pointer to the buffer into a channel

shred processor thread:

10. Read the pointer from the channel
11. Write the shreds into the blockstore

With this approach, steps 3, 4, 6, 7, 9, 10, and 11 each copy 120 KB. At best with copy elision, you might see something like only steps 3, 6, 9, and 11 each copy 120 KB. So the amount of data copied is 2-3.5 times as much as the arraylist approach.

Reply:
While I can't disagree that there will be more copies,

1. These copies are tens of times faster than the heap allocation that the array list would be doing.
2. We have much more important optimization to do before we need to worry about doing too many copies.

It may seem like copying the packets is bad for performance, however, when the alternative is our current excessive usage of array lists and heap allocation, it wins.

Reply:
Fair point about the allocations. One way to deal with this is by recycling them.

The shred collector had sufficiently good performance before this change to process many more packets than mainnet produces. Obviously we'd like this to be as fast as possible. I assume that you have improved the performance with this change. But since there's a tradeoff, it's hard to be sure without some tests or benchmarks to demonstrate it.

Reply:
I'm satisfied with the allocation argument, so I would only want it to pass a basic sanity check in terms of performance. The shred collector has never been formally benchmarked. The criteria for "good enough" so far has only been "Can I run it on my t3.xlarge instance in ec2, and it catches up to mainnet from at least 100 slots behind, and doesn't fall behind after running for a few minutes?"

Reply:
I'm happy to do all of these, could you explain the steps? I agree with @0xNineteen that we should try to get this PR through soon, it's quite large and fixes a lot of bugs.


I'm also happy to do formal benchmarking of the shred collector, but after I get through the snapshot benchmarking, continuous benchmarking with the ec2, and getting telemetry for historical data. Just have a bit on the plate already :).

Reply:
@dnut would you like me to do some benchmarking, or can this be done as a follow up? Need this convo resolved to merge 😄.